### DISCUSSION TITLE
Confusing metric result for used languages

### DISCUSSION THREAD

**Discussion by ttytm** [lowlighter/metrics#1047] [Q&A] [answered]
Some issues I experience are:

- Indepth analysis doesn't contain some languages used in public repos and none of those used in private repos while  normal analysis does.
  - E.g. for private: I'm having some js and svelte projects as private repos which are captured when indepth analysis is turned off
  - public: I'm doing some rust programming in a public repo on my account, which is captured **without** indepth analysis but left out when it's turned on. Rust also isn't captured by the recent languages metric (The dominant lang in the repo i'm using rust is typescript, could that be the reason? But still, why is it captured in a normal analysis)?
 
<details>
<summary>Recent metrics that resemble the issue described above of normal vs indepth metrics. <em>Click to unfold images</em></summary>

<img src="https://user-images.githubusercontent.com/34311583/167621608-4f008f5f-b796-4c08-b7b2-7fdc63d52107.svg" alt="Metrics" width="400">

<img src="https://user-images.githubusercontent.com/34311583/167621601-59f94357-3b67-4067-8646-2dc459595c0a.svg" alt="Metrics" width="400">

</details>

- What is being analyzed? E.g. with deletion of repositories 
  - Until recently, I had some repositories that I forked and contributed upstream via PRs. To clean up I deleted them because my work was done there but deleting them caused the used languages to disappear (e.g. Python on the screenshots isn't in my used langs anymore), even though the PRs were accepted and the upstream repo contains the commits made from the account I'm using.

Reading the docs and trying out different analysis methods I couldn't figure out why this is and left confused.  For now, I'm thinking that those metrics are a cool gimmick but not a representative insight into metrics at all. But that's probably me misunderstanding the workings. So if it's possible for you to leave some words of clarification and how to use this tool in a better way that would be great. 

Edit: I'm using an action with a secret token and gave all necessary rights

**Comment by spenserblack** [comment]
For the lack of Rust detection, it's most likely related to https://github.com/lowlighter/metrics/issues/551 (see https://github.com/lowlighter/metrics/issues/551#issuecomment-948570921 for my theory why this is happening, but it's basically a combination of `.rs` defaulting to RenderScript and not Rust when there's not enough code for heuristics).

**Reply by ttytm** [reply]
> For the lack of Rust detection, it's most likely related to #551 (see [#551 (comment)](https://github.com/lowlighter/metrics/issues/551#issuecomment-948570921) for my theory why this is happening, but it's basically a combination of `.rs` defaulting to RenderScript and not Rust when there's not enough code for heuristics).

Ahh I see, seems obvious when rust disappeares and renderscript appears. So that's probably it for the issue with rust code.

**Comment by lowlighter** [comment, marked as answer]
The aim behind `indepth` mode is mostly explained [here](https://github.com/lowlighter/metrics/blob/master/source/plugins/languages/README.md#-indepth-mode) but let me know if it isn't clear enough

But to recap quickly: 
- Default algorithm (that most tools also use AFAIK) is to fetch the "languages recap" (which is also displayed in side bar of repo on GitHub) from each repo and basically sum it up
  - **Pros**
    - Quick and fast
    - Works great if you're sole contributor
  - **Cons**
    - Less representative on collaborative projects (you may not have actually authored the whole codebase)
    - Leave out any code written in the past since only "current state of repositories" is taken into account
- Indepth algorithm clones repositories and filter out the commit you've authored, and check each patch from them before summing them up
  - **Pros**
    - Should be more accurate, only code you've authored are considered "yours"
    - Iterate through all git history, and considers any refactor/editions as written code, plus it counts both lines of codes and bytes
  - **Cons**
    - Slow, may not work on huge repositories or if you need to analyze a lot of repositories 
    - [`commits_authoring`](https://github.com/lowlighter/metrics/blob/master/source/plugins/base/README.md#commits_authoring) needs to be properly configured to accurately determine ownership

> Indepth analysis doesn't contain some languages used in public repos and none of those used in private repos while normal analysis does.

So as stated above, ideally if a public repo contains let's say rust and typescript and you've worked only on rust code, the end goal is to "leave out" typescript from your stats since you didn't really author it.

As for private repos, another user reported this issue too, it may be an actual bug but I haven't investigated too much on it yet

> public: I'm doing some rust programming in a public repo on my account, which is captured without indepth analysis but left out when it's turned on. Rust also isn't captured by the recent languages metric (The dominant lang in the repo i'm using rust is typescript, could that be the reason? But still, why is it captured in a normal analysis)?

Most likely @spenserblack is correct.

Indepth mode should not have the issue though (so needs to investigate on this too), but for recently used languages it's kind of normal.

The way recently used language works is to fetch "git patches" from your activity feed, it doesn't actually look at the whole file or repository, it just perform the languages analysis on code snippet so it may be lacking the required context so github linguist is able to accurately determine the used language. Maybe a workaround would be to load the entire file to get the languages and then evaluate the patch

I'm not sure the linguist api provides it, but maybe if we could pass "weights" to heuristics it could mitigate the issue too.

> What is being analyzed? E.g. with deletion of repositories ....

This one is bit tricky, but only repositories that were fetched by the base plugin (which depends on [`repositories_affiliations`](https://github.com/lowlighter/metrics/blob/master/source/plugins/base/README.md#repositories_affiliations) and [`repositories_forks`](https://github.com/lowlighter/metrics/blob/master/source/plugins/base/README.md#repositories_affiliations)) are analyzed.

If you let the default settings, then only repositories that you own (and are not forks) are analyzed.
There's also some work to manually include additional repositories through https://github.com/lowlighter/metrics/issues/857 but it hasn't been done yet.

___

Hopefully this is a bit more clear and it's helpful 😅

**Comment by ttytm** [comment]
Fantasitc reply Mr. 

makes it already more clear then I had expected. Thanks alot.
I'm looking forword to the developments happening!

**Reply by lowlighter** [reply]
Thanks a lot for your support :heart:  !
